go/vt/vtgate/grpcvtgateservice/server.go

```diff
@@ -187,12 +241,15 @@ func (vtg *VTGate) StreamExecute(request *vtgatepb.StreamExecuteRequest, stream
 		session = &vtgatepb.Session{Autocommit: true}
 	}
 
+	ss := newStreamSender()
+	ss.start(stream)
+	defer ss.close()
+
 	session, vtgErr := vtg.server.StreamExecute(ctx, nil, session, request.Query.Sql, request.Query.BindVariables, func(value *sqltypes.Result) error {
-		// Send is not safe to call concurrently, but vtgate
-		// guarantees that it's not.
-		return stream.Send(&vtgatepb.StreamExecuteResponse{
+		resp := &vtgatepb.StreamExecuteResponse{
 			Result: sqltypes.ResultToProto3(value),
-		})
+		}
+		return ss.sendMessage(resp)
```

Comment:
vtgate always sends a single stream of responses. It is protected by the mutex in `executor` which is above any engine primitive. So, we do not need this streaming protection here.